realm/realm-library/src/main/cpp/jni_util/log.hpp

```diff
@@ -0,0 +1,193 @@
+    inline static void i(const char* message)
+    {
+        shared().log(error, REALM_JNI_TAG, nullptr, nullptr, message);
+    }
+    inline static void w(const char* message)
+    {
+        shared().log(error, REALM_JNI_TAG, nullptr, nullptr, message);
+    }
+    inline static void e(const char* message)
+    {
+        shared().log(error, REALM_JNI_TAG, nullptr, nullptr, message);
+    }
+    inline static void f(const char* message)
+    {
+        shared().log(error, REALM_JNI_TAG, nullptr, nullptr, message);
+    }
+
+    template<typename... Args>
+    inline static void t(const char* fmt, Args&&... args)
+    {
+        shared().log(trace, REALM_JNI_TAG, nullptr, nullptr, _impl::format(fmt, {_impl::Printable(args)...}).c_str());
+    }
+    template<typename... Args>
+    inline static void d(const char* fmt, Args&&... args)
+    {
+        shared().log(debug, REALM_JNI_TAG, nullptr, nullptr, _impl::format(fmt, {_impl::Printable(args)...}).c_str());
+    }
+    template<typename... Args>
+    inline static void i(const char* fmt, Args&&... args)
+    {
+        shared().log(info, REALM_JNI_TAG, nullptr, nullptr, _impl::format(fmt, {_impl::Printable(args)...}).c_str());
+    }
+    template<typename... Args>
+    inline static void w(const char* fmt, Args&&... args)
+    {
+        shared().log(warn, REALM_JNI_TAG, nullptr, nullptr, _impl::format(fmt, {_impl::Printable(args)...}).c_str());
+    }
+    template<typename... Args>
+    inline static void e(const char* fmt, Args&&... args)
+    {
+        shared().log(error, REALM_JNI_TAG, nullptr, nullptr, _impl::format(fmt, {_impl::Printable(args)...}).c_str());
+    }
+    template<typename... Args>
+    inline static void f(const char* fmt, Args&&... args) {
+        shared().log(fatal, REALM_JNI_TAG, nullptr, nullptr, _impl::format(fmt, {_impl::Printable(args)...}).c_str());
+    }
+
+    // Get the shared Log instance.
+    static Log& shared();
+
+    // public & static for reading faster. For TR_ENTER check.
+    // Accessing to this var won't be thread safe and it is not necessary to be. Changing log level concurrently
+    // won't be a critical issue for commons cases.
+    static Level s_level;
+private:
+    Log();
+
+    std::vector<std::shared_ptr<JniLogger>> m_loggers;
+    std::mutex m_mutex;
+    // Log tag for generic Realm JNI.
+    static const char* REALM_JNI_TAG;
+
+};
+
+// Base Logger class.
+class JniLogger {
+protected:
+    JniLogger();
+    // Used by JavaLogger.
+    JniLogger(bool is_java_logger);
+    // Indicate if this is a wrapper for Java Logger class. See JavaLogger
+    bool m_is_java_logger;
+
+protected:
+    // Overwrite this method to handle the log event.
+    // throwable is the Throwable passed from Java which could be null.
+    // statcktrace is the stracktrace string parsed from throwable by the Java layer which could be an empty string.
+    virtual void log(Log::Level level, const char* tag, jthrowable throwable, const char* stacktrace,
```

Comment:
I think we need to parse the Throwable to custom loggers, yes. Especially since they are prime target for more validation/formatting in a custom logger.